src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+        Calculate CL rebase from prev_blockstamp to ref_blockstamp given the changes in withdrawal vault
+        """
+        curr_ref_epoch = curr_ref_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        prev_epoch = prev_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        logger.info({"msg": f"Calculating CL rebase between {prev_epoch, curr_ref_epoch=}"})
+
+        ref_lido_balance_with_vault = sum(self._calculate_lido_balance_with_vault(
+            self.lido_validators, curr_ref_blockstamp
+        ))
+
+        prev_all_validators = self.w3.cc.get_validators(prev_blockstamp.state_root)
+        prev_lido_validators = self.w3.lido_validators.filter_lido_validators(self.lido_keys, prev_all_validators)
+        prev_lido_balance_with_vault = sum(self._calculate_lido_balance_with_vault(
+            prev_lido_validators, prev_blockstamp
+        ))
+
+        # handle 32 ETH balances of freshly baked validators, who was activated between epochs
+        validators_diff_in_gwei = (len(self.lido_validators) - len(prev_lido_validators)) * MIN_DEPOSIT_AMOUNT
+        assert validators_diff_in_gwei >= 0, "Validators diff should be positive or 0. Something went wrong with CL API"
+
+        corrected_prev_lido_balance_with_vault = (
+            prev_lido_balance_with_vault
+            + validators_diff_in_gwei
+            - self._get_withdrawn_from_vault_between(prev_blockstamp, curr_ref_blockstamp)
+        )
+
+        cl_rebase = ref_lido_balance_with_vault - corrected_prev_lido_balance_with_vault
+
+        logger.info({"msg": f"CL rebase from {prev_epoch=}: {cl_rebase} Gwei"})
+
+        return cl_rebase
+
+    def _calculate_lido_balance_with_vault(
+        self, lido_validators: list[LidoValidator], vault_balance_at_blockstamp: BlockStamp
+    ) -> tuple[Gwei, Gwei]:
+        """
+        Calculate Lido validators balance and balance in withdrawal vault contract in particular block
+        """
+        lido_balance = sum(int(v.validator.balance) for v in lido_validators)
+        withdrawal_vault_balance = self.w3.eth.get_balance(
+            self.w3.lido_contracts.withdrawal_vault.address, vault_balance_at_blockstamp.block_hash
+        )
+
+        return Gwei(lido_balance), Gwei(self.w3.from_wei(withdrawal_vault_balance, 'gwei'))
+
+    def _get_withdrawn_from_vault_between(self, prev_blockstamp: BlockStamp, curr_blockstamp: BlockStamp) -> int:
+        """
+        Lookup for ETHDistributed event and sum up all withdrawalsWithdrawn
+        """
+        prev_epoch = prev_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        curr_epoch = curr_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        logger.info({"msg": f"Get withdrawn from vault between {prev_epoch,curr_epoch=}"})
+        vault_withdrawals = 0
+        events = self.w3.lido_contracts.lido.events.ETHDistributed.get_logs(
+            from_block=int(prev_blockstamp.block_number) + 1, to_block=int(curr_blockstamp.block_number)
```

Comment:
Could you add a comment on why we are adding 1 block here?

Reply:
done